theano/tensor/opt.py

```diff
@@ -6223,6 +6223,33 @@ def add_calculate(num, denum, aslist=False, out_type=None):
 register_canonicalize(local_add_canonizer, name='local_add_canonizer')
 
 
+@register_canonicalize
+@gof.local_optimizer([T.mul])
+def local_mul_exp(node):
+    if not node.op == T.mul:
+        return False
+
+    pow_base_dict = defaultdict(list)
+    unchanged_inputs = []
+    pow_nodes_cnt = 0
+    for i in node.inputs:
+        if i.owner is not None and isinstance(i.owner.op, T.Elemwise) and\
+           i.owner.op.scalar_op in [T.pow, theano.scalar.basic.pow]:
+            assert len(i.owner.inputs) == 2
+            base = i.owner.inputs[0]
+            pow_base_dict[base].append(i.owner.inputs[1])
+            pow_nodes_cnt += 1
+        else:
+            unchanged_inputs.append(i)
+    if pow_nodes_cnt > 1:
+        mul_list = unchanged_inputs
+        for base, exp in iteritems(pow_base_dict):
+            mul_list.append(T.pow(base, T.add(*exp)))
```

Comment:
This is good, but would request extra work by the optimizer to clean up some corner cases.

If exp is a list of only 1 element, the add will be useless and will be removed by later optimization pass. I would prefer that we don't introduce it in the first place. Can you do something like:
```
if len(exp) > 1:
    tmp = T.add(*exp)
else:
    tmp = exp[0]
T.pow(base, tmp)
```